Add support for HTTPFallback

sysrqb opened the pull request:
#### 6660f0c6e9c9e6e921cbcf7d971a9fae05f14541
<pre>
Add support for HTTPFallback
<a href="https://bugs.webkit.org/show_bug.cgi?id=248668">https://bugs.webkit.org/show_bug.cgi?id=248668</a>
rdar://problem/102911054

Reviewed by Chris Dumez.

When loading a frame results in a failure because of an error related to
establishing a TLS connection, then we may be able to successfully request the
resource as HTTP. We attempt this automatic fallback behavior when the
HTTPSFirst policy is active and we are not currently attempting HTTPFallback
and the request was not itself the result of a HTTP redirect.
For example, an allowed fallback case would be:

<a href="https://site1.example">https://site1.example</a> doesn&apos;t support TLS, so we fallback to
<a href="http://site1.example">http://site1.example</a>

An unsupported fallback case is:

<a href="https://site1.example">https://site1.example</a> supports TLS and redirects to <a href="https://site2.example">https://site2.example</a>, but
<a href="https://site2.example">https://site2.example</a> doesn&apos;t support TLS and we fallback to
<a href="http://site1.example.">http://site1.example.</a> Instead, the load fails on <a href="https://site2.example.">https://site2.example.</a>

The delegate methods didFailProvisionalLoadForFrame and
didFailProvisionalNavigationWithError are not called when we follow this
fallback behavior.

Covered by API tests.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::dispatchDidFailProvisionalLoad):
(WebCore::FrameLoader::handleLoadFailureRecovery):
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/FrameLoaderClient.h:
* Source/WebCore/loader/FrameLoaderTypes.h:
* Source/WebCore/platform/network/ResourceErrorBase.h:
* Source/WebCore/platform/network/cf/ResourceError.h:
* Source/WebCore/platform/network/curl/ResourceError.h:
(WebCore::ResourceError::errorRecoveryMethod const):
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::ResourceError::errorRecoveryMethod const):
* Source/WebCore/platform/network/soup/ResourceError.h:
(WebCore::ResourceError::errorRecoveryMethod const):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::cancel):
(WebKit::ProvisionalPageProxy::didFailProvisionalLoadForFrame):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFailProvisionalLoadForFrame):
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebKitLegacy/win/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebKitLegacy/win/WebCoreSupport/WebFrameLoaderClient.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/260213@main">https://commits.webkit.org/260213@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69fd757702a2cdeb48c785a7cca22a50694d3a98

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105903 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14971 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38749 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115097 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114853 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109805 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16389 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6165 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98131 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111653 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12483 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95447 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40023 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/109354 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94348 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27091 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81692 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8236 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28445 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8717 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5030 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14347 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47985 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7061 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10272 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/network/ResourceErrorBase.h:
```diff
@@ -58,6 +58,11 @@ class ResourceErrorBase {
     };
     enum class IsSanitized : bool { No, Yes };
 
+    enum class ErrorRecoveryMethod : uint8_t {
```
`: bool` ?

sysrqb replied:
I used `uint8_t` for extensibility, but you're right that we can just make this a bool right now.

cdumez commented on Source/WebCore/platform/network/ResourceErrorBase.h:
```diff
@@ -124,4 +133,12 @@ template<> struct EnumTraits<WebCore::ResourceErrorBase::Type> {
     >;
 };
 
+template<> struct EnumTraits<WebCore::ResourceErrorBase::ErrorRecoveryMethod> {
```
Wouldn't be needed if you had used `: bool` I believe.

cdumez reviewed (approved):
LGTM

achristensen07 commented on Source/WebCore/platform/network/ResourceErrorBase.h:
```diff
@@ -99,6 +107,7 @@ class ResourceErrorBase {
     int m_errorCode { 0 };
     Type m_type { Type::General };
     IsSanitized m_isSanitized { IsSanitized::No };
+    ErrorRecoveryMethod m_errorRecoveryMethod { ErrorRecoveryMethod::NoRecovery };
```
We don't need additional memory to store this, and please don't add it.  We already store the domain and error code from which this can be derived.

achristensen07 replied:
Add just a getter, and the non-cocoa version should probably always return ErrorRecoveryMethod::NoRecovery

sysrqb replied:
Good point, changed.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2555,6 +2555,27 @@ void FrameLoader::dispatchDidFailProvisionalLoad(DocumentLoader& provisionalDocu
+    FRAMELOADER_RELEASE_LOG(ResourceLoading, "handleLoadFailureRecovery: errorRecoveryMethod: %hhu, isHTTPSFirstApplicable: %d, isHTTPFallbackInProgress: %d", error.errorRecoveryMethod(), isHTTPSFirstApplicable, isHTTPFallbackInProgress());
+    if (error.errorRecoveryMethod() == ResourceError::ErrorRecoveryMethod::NoRecovery || !isHTTPSFirstApplicable || isHTTPFallbackInProgress()) {
+        if (isHTTPFallbackInProgress())
+            setHTTPFallbackInProgress(true);
```
This doesn't appear to do anything.  Did you mean false?

sysrqb replied:
Yes, thanks for catching that.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1098,3 +1099,111 @@ HTTPServer server({
+    EXPECT_EQ(errorCode, -1202);
+    EXPECT_FALSE(finishedSuccessfully);
+    EXPECT_EQ(loadCount, 3);
+}
```
Could you add a test to check if window.location is http or https after successfully downgrading?
Could you add a test to check that something that would use Document::isSecureContext() fails after successfully downgrading?

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1098,3 +1099,111 @@ HTTPServer server({
+    while (!errorCode)
+        TestWebKitAPI::Util::spinRunLoop(10);
+
+    EXPECT_EQ(errorCode, -1202);
```
Please use NSURLErrorServerCertificateUntrusted

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1098,3 +1099,111 @@ HTTPServer server({
+    while (!errorCode)
+        TestWebKitAPI::Util::spinRunLoop(10);
+
+    EXPECT_EQ(errorCode, -1202);
```
Ditto

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1098,3 +1099,111 @@ HTTPServer server({
+{
+    using namespace TestWebKitAPI;
+    HTTPServer httpsServer({
+        { "/"_s, { { }, "hi"_s } }
```
I like to add something to the path to avoid confusion and unexpected surprises, like "/secure"

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1098,3 +1099,111 @@ HTTPServer server({
+    [webView setNavigationDelegate:delegate.get()];
+    [webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://site.example"]]];
+    while (!errorCode)
+        TestWebKitAPI::Util::spinRunLoop(10);
```
10 seems excessive, but it's not really important.

mcatanzaro commented on Source/WebCore/platform/network/soup/ResourceError.h:
```diff
@@ -60,6 +60,8 @@ class ResourceError : public ResourceErrorBase {
     GTlsCertificate* certificate() const { return m_certificate.get(); }
     void setCertificate(GTlsCertificate* certificate) { m_certificate = certificate; }
 
+    ErrorRecoveryMethod errorRecoveryMethod() const { return ErrorRecoveryMethod::NoRecovery; }
```
This looks really cool. So to implement HTTPS First or HTTPS Only mode for GTK/WPE, we would need to implement this function here, plus add an API to set the network connection integrity policy?

How did you decide which errors should result in HTTPFallback vs. which errors should not? All errors here are surely going to be network errors, and it seems you're handling errors at IP layer, TLS layer, and even HTTP layer, so it's a little unclear to me what sorts of errors should result in fallback and which should not.

sysrqb replied:
Yes! That's correct, this will require enabling the feature via the NetworkConnectionIntegrity option. The relevant error cases are somewhat a judgement call. This patch aims for falling back on the specific platform for cases where the endpoint either doesn't support HTTPS/TLS/QUIC or that endpoint web server is probably misconfigured. This assumes that the original request was valid/correctly formed, but of course that assumption could be wrong.

sysrqb committed:
Add support for HTTPFallback
https://bugs.webkit.org/show_bug.cgi?id=248668
rdar://problem/102911054

Reviewed by Chris Dumez.

When loading a frame results in a failure because of an error related to
establishing a TLS connection, then we may be able to successfully request the
resource as HTTP. We attempt this automatic fallback behavior when the
HTTPSFirst policy is active and we are not currently attempting HTTPFallback
and the request was not itself the result of a HTTP redirect.
For example, an allowed fallback case would be:

https://site1.example doesn't support TLS, so we fallback to
http://site1.example

An unsupported fallback case is:

https://site1.example supports TLS and redirects to https://site2.example, but
https://site2.example doesn't support TLS and we fallback to
http://site1.example. Instead, the load fails on https://site2.example.

The delegate methods didFailProvisionalLoadForFrame and
didFailProvisionalNavigationWithError are not called when we follow this
fallback behavior.

Covered by API tests.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::dispatchDidFailProvisionalLoad):
(WebCore::FrameLoader::handleLoadFailureRecovery):
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/FrameLoaderClient.h:
* Source/WebCore/loader/FrameLoaderTypes.h:
* Source/WebCore/platform/network/ResourceErrorBase.h:
* Source/WebCore/platform/network/cf/ResourceError.h:
* Source/WebCore/platform/network/curl/ResourceError.h:
(WebCore::ResourceError::errorRecoveryMethod const):
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::ResourceError::errorRecoveryMethod const):
* Source/WebCore/platform/network/soup/ResourceError.h:
(WebCore::ResourceError::errorRecoveryMethod const):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::cancel):
(WebKit::ProvisionalPageProxy::didFailProvisionalLoadForFrame):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFailProvisionalLoadForFrame):
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebKitLegacy/win/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebKitLegacy/win/WebCoreSupport/WebFrameLoaderClient.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST):

Canonical link: https://commits.webkit.org/260213@main
